### PR TITLE
Spark: Fix issue when partitioning by UUID

### LINKED ISSUE #8247: Spark: Support UUID partitioned tables
### Feature Request / Improvement

If you create an UUID table with a partition:

```python
from pyiceberg.catalog import load_catalog
from pyiceberg.schema import Schema
from pyiceberg.types import UUIDType, NestedField

cat = load_catalog(
    "prod",
    **properties
)

schema = Schema(
    NestedField(
        field_id=1, name="c1", field_type=UUIDType(), required=False
    ),
)

from pyiceberg.partitioning import PartitionSpec, PartitionField
from pyiceberg.transforms import BucketTransform

partition_spec = PartitionSpec(
    PartitionField(
        source_id=1, field_id=1000, transform=BucketTransform(16), name="c1_by_bucket"
    )
)

cat.create_table(
    identifier="default.uuid",
    schema=schema,
    partition_spec=partition_spec,
)
```

And you try to write to it, you'll get:

```
spark-sql (default)> insert into uuid VALUES('102cb62f-e6f8-4eb0-9973-d9b012ff0967');
23/08/07 09:20:31 ERROR Utils: Aborting task
java.lang.ClassCastException: Cannot cast java.lang.String to java.util.UUID
	at java.base/java.lang.Class.cast(Class.java:3605)
	at org.apache.iceberg.spark.source.InternalRowWrapper.get(InternalRowWrapper.java:63)
	at org.apache.iceberg.Accessors$PositionAccessor.get(Accessors.java:71)
	at org.apache.iceberg.Accessors$PositionAccessor.get(Accessors.java:58)
	at org.apache.iceberg.PartitionKey.partition(PartitionKey.java:106)
	at org.apache.iceberg.spark.source.SparkWrite$PartitionedDataWriter.write(SparkWrite.java:752)
	at org.apache.iceberg.spark.source.SparkWrite$PartitionedDataWriter.write(SparkWrite.java:723)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.write(WriteToDataSourceV2Exec.scala:516)
	at org.apache.spark.sql.execution.datasources.v2.WritingSparkTask.$anonfun$run$1(WriteToDataSourceV2Exec.scala:471)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1563)
	at org.apache.spark.sql.execution.datasources.v2.WritingSparkTask.run(WriteToDataSourceV2Exec.scala:509)
	at org.apache.spark.sql.execution.datasources.v2.WritingSparkTask.run$(WriteToDataSourceV2Exec.scala:448)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.run(WriteToDataSourceV2Exec.scala:514)
	at org.apache.spark.sql.execution.datasources.v2.V2TableWriteExec.$anonfun$writeWithV2$2(WriteToDataSourceV2Exec.scala:411)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:92)
	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:161)
	at org.apache.spark.scheduler.Task.run(Task.scala:139)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:554)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1529)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:557)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
23/08/07 09:20:31 ERROR DataWritingSparkTask: Aborting commit for partition 0 (task 1, attempt 0, stage 2.0)
23/08/07 09:20:31 ERROR DataWritingSparkTask: Aborted commit for partition 0 (task 1, attempt 0, stage 2.0)
23/08/07 09:20:31 ERROR Executor: Exception in task 0.0 in stage 2.0 (TID 1)
java.lang.ClassCastException: Cannot cast java.lang.String to java.util.UUID
	at java.base/java.lang.Class.cast(Class.java:3605)
	at org.apache.iceberg.spark.source.InternalRowWrapper.get(InternalRowWrapper.java:63)
	at org.apache.iceberg.Accessors$PositionAccessor.get(Accessors.java:71)
	at org.apache.iceberg.Accessors$PositionAccessor.get(Accessors.java:58)
	at org.apache.iceberg.PartitionKey.partition(PartitionKey.java:106)
	at org.apache.iceberg.spark.source.SparkWrite$PartitionedDataWriter.write(SparkWrite.java:752)
	at org.apache.iceberg.spark.source.SparkWrite$PartitionedDataWriter.write(SparkWrite.java:723)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.write(WriteToDataSourceV2Exec.scala:516)
	at org.apache.spark.sql.execution.datasources.v2.WritingSparkTask.$anonfun$run$1(WriteToDataSourceV2Exec.scala:471)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1563)
	at org.apache.spark.sql.execution.datasources.v2.WritingSparkTask.run(WriteToDataSourceV2Exec.scala:509)
	at org.apache.spark.sql.execution.datasources.v2.WritingSparkTask.run$(WriteToDataSourceV2Exec.scala:448)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.run(WriteToDataSourceV2Exec.scala:514)
	at org.apache.spark.sql.execution.datasources.v2.V2TableWriteExec.$anonfun$writeWithV2$2(WriteToDataSourceV2Exec.scala:411)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:92)
	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:161)
	at org.apache.spark.scheduler.Task.run(Task.scala:139)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:554)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1529)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:557)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
23/08/07 09:20:31 WARN TaskSetManager: Lost task 0.0 in stage 2.0 (TID 1) (192.168.1.61 executor driver): java.lang.ClassCastException: Cannot cast java.lang.String to java.util.UUID
	at java.base/java.lang.Class.cast(Class.java:3605)
	at org.apache.iceberg.spark.source.InternalRowWrapper.get(InternalRowWrapper.java:63)
	at org.apache.iceberg.Accessors$PositionAccessor.get(Accessors.java:71)
	at org.apache.iceberg.Accessors$PositionAccessor.get(Accessors.java:58)
	at org.apache.iceberg.PartitionKey.partition(PartitionKey.java:106)
	at org.apache.iceberg.spark.source.SparkWrite$PartitionedDataWriter.write(SparkWrite.java:752)
	at org.apache.iceberg.spark.source.SparkWrite$PartitionedDataWriter.write(SparkWrite.java:723)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.write(WriteToDataSourceV2Exec.scala:516)
	at org.apache.spark.sql.execution.datasources.v2.WritingSparkTask.$anonfun$run$1(WriteToDataSourceV2Exec.scala:471)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1563)
	at org.apache.spark.sql.execution.datasources.v2.WritingSparkTask.run(WriteToDataSourceV2Exec.scala:509)
	at org.apache.spark.sql.execution.datasources.v2.WritingSparkTask.run$(WriteToDataSourceV2Exec.scala:448)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.run(WriteToDataSourceV2Exec.scala:514)
	at org.apache.spark.sql.execution.datasources.v2.V2TableWriteExec.$anonfun$writeWithV2$2(WriteToDataSourceV2Exec.scala:411)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:92)
	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:161)
	at org.apache.spark.scheduler.Task.run(Task.scala:139)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:554)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1529)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:557)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

Tables without a UUID partition work great

### Query engine

None

### PR DESCRIPTION
**Pull request by nastra**
fixes #8247

The isse (as can be seen in #8247) is that `InternalRowWrapper` uses the underlying Java class of a particular type (`UUID.class` in this case) to do a cast. Since Spark doesn't support UUIDs, they are mapped as strings and the cast fails. 

To solve the issue, we're passing the underlying Iceberg schema as struct to `InternalRowWrapper`, so that we can convert to the right type before doing the cast. A similar approach is being done for Flink in `RowDataWrapper`.

Unfortunately this required to fix this issue across all Spark versions, because `RecordWrapperTest` is being used by all Spark versions. The fix itself is inside `InternalRowWrapper` and all other changes are due to passing the underlying Schema as struct to `InternalRowWrapper`.

### CODE REVIEW

**Comment by manuzhang** [comment]
Which issue is this fixing?

File: data/src/test/java/org/apache/iceberg/RecordWrapperTest.java [hunk 1]
```diff
@@ -44,8 +44,8 @@ public abstract class RecordWrapperTest {
           optional(113, "bytes", Types.BinaryType.get()),
           required(114, "dec_9_0", Types.DecimalType.of(9, 0)),
           required(115, "dec_11_2", Types.DecimalType.of(11, 2)),
-          required(116, "dec_38_10", Types.DecimalType.of(38, 10)) // maximum precision
-          );
+          required(116, "dec_38_10", Types.DecimalType.of(38, 10)), // maximum precision
+          optional(117, "uuid", Types.UUIDType.get()));
```
Comment (nastra) [inline data/src/test/java/org/apache/iceberg/RecordWrapperTest.java:48]: this is being used by Spark tests across all versions, so we need to apply the fix for all Spark versions in a single PR unfortunately

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/InternalRowWrapper.java [hunk 2]
```diff
@@ -71,8 +77,12 @@ public <T> void set(int pos, T value) {
     row.update(pos, value);
   }
 
-  private static BiFunction<InternalRow, Integer, ?> getter(DataType type) {
+  private static BiFunction<InternalRow, Integer, ?> getter(Type icebergType, DataType type) {
```
Comment (hililiwei) [inline spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/InternalRowWrapper.java:80]: nit: Do we rename the `type` to make it clear that it is `sql type` of spark?

**Comment by nastra** [comment]
@aokolnychyi could you take a look at this please?

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/InternalRowWrapper.java [hunk 3]
```diff
@@ -40,9 +43,12 @@ class InternalRowWrapper implements StructLike {
   private InternalRow row = null;
 
   @SuppressWarnings("unchecked")
-  InternalRowWrapper(StructType rowType) {
+  InternalRowWrapper(StructType rowType, Types.StructType icebergStruct) {
```
Comment (singhpk234) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/InternalRowWrapper.java:46]: should we have a Pre-condition check that icebergStruct.fields().length is equal to types.lenght ?

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/InternalRowWrapper.java [hunk 4]
```diff
@@ -71,8 +77,12 @@ public <T> void set(int pos, T value) {
     row.update(pos, value);
   }
 
-  private static BiFunction<InternalRow, Integer, ?> getter(DataType type) {
+  private static BiFunction<InternalRow, Integer, ?> getter(Type icebergType, DataType type) {
     if (type instanceof StringType) {
+      if (Type.TypeID.UUID == icebergType.typeId()) {
```
Comment (singhpk234) [inline spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/InternalRowWrapper.java:82]: can we add a comment here explaining that a StringType may be a UUID-type hence return UUID also should we explicity check the else case saying if it's not StringType either throw UnReachable or Unsupported ops or null accordingly ?

File: spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/InternalRowWrapper.java (hunk 2 above)
Comment (nastra) [reply spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/InternalRowWrapper.java:80]: I'd probably just keep it `type` to cause less changes

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/InternalRowWrapper.java (hunk 3 above)
Comment (nastra) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/InternalRowWrapper.java:46]: I've added one

**Review by singhpk234 — APPROVED** [review]
LGTM, Thanks @nastra !

File: spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/InternalRowWrapper.java [hunk 5]
```diff
@@ -40,9 +44,17 @@ class InternalRowWrapper implements StructLike {
   private InternalRow row = null;
 
   @SuppressWarnings("unchecked")
-  InternalRowWrapper(StructType rowType) {
+  InternalRowWrapper(StructType rowType, Types.StructType icebergStruct) {
```
[the remark this answers is not in the document]
Comment (nastra) [reply spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/InternalRowWrapper.java]: makes sense, I've updated that

**Comment by nastra** [comment]
thanks for the reviews @hililiwei, @singhpk234, @amogh-jahagirdar

**Pull request merged**